### ISSUE TITLE
3.1.6 png transparency shows black in wxTreeCtrl

### ISSUE THREAD

**Issue by msbroadf** [wxWidgets/wxWidgets#22349]
My app contains a wxTreeCtrl with png images used and these have transparency. 

In 3.1.4 this displayed correctly. Recompiling with 3.1.6 with no other changes to my code shows the same icons in the wxTreeCtrl but with black instead of the transparent area. 

Is this something related to the wxBitmapBundle not correctly supporting transparency at the moment?

**Comment by vadz** [comment]
This is not supposed to happen, so it looks like a bug but, as always, it would be great to have a simple way of reproducing this, ideally in the treectrl sample. Could you please make a minimal patch to it showing the problem?

Also, which platform does this happen under?

**Comment by vadz** [comment]
It would be great to at least have the PNG files showing the problem.

**Comment by msbroadf** [comment]
Sorry i have been trying to compile the treectl example but havent yet succeeded (my build process is very complicated) but i have attached the png with the issue for the time being

![audio](https://user-images.githubusercontent.com/13748102/165187153-cdb80e36-347a-49eb-835b-42d4242a0504.png)

**Comment by vadz** [comment]
Thanks, I'll test with this one, but normally you should be able to just use the provided make- or project file to build the sample without any changes.

**Comment by PBfordev** [comment]
I cannot confirm this on MSW. I used the current master but I do not see any relevant changes since 3.1.6.

I downloaded the image provided by msbroadf, used bundled bin2c to embed it in the tree sample and use it instead of "file" icon there. I also set the background color to yellow, to make the transparency stand out.
![treectrl-sample-png](https://user-images.githubusercontent.com/12495521/165263276-73c0da37-ecf2-4986-ac6f-416e7d998f2c.png)

The patch is here: https://github.com/PBfordev/wxWidgets/commit/7b30db394238dfb2eff75ec945a6d8534284de7c

**Comment by msbroadf** [comment]
I managed to compile with no changes to treectrl example and it looks like this:

![Untitled](https://user-images.githubusercontent.com/13748102/165270923-51b7deac-e218-4c18-9b82-ff054f7c98de.png)

My setup is as follows:

wxWidgets-3.1.6.zip downloaded from github
mingw64 cross compiler building exe on linux

**Comment by PBfordev** [comment]
@msbroadf, It seems that your sample is missing an application manifest: Not only is it not DPI-aware (blurry text), the control borders (see e.g. the status bar) have 3-D effect betraying Win95 look. Did you build the sample using the unmodified bundled makefile? If so then it is very odd, it should have the manifest embedded including PMv2 DPI-awareness.

Anyway, I deleted the application manifest from my sample executable and ran it also on Win11 (inside a VM). Transparency still worked on both Windows 10 and 11.

I also tried the same with MinGW build (MSYS2 package mingw-w64-x86_64-toolchain, GCC 11.2). The transparency in the original sample worked. However, when I deleted the application manifest, the transparent pixels were displayed black, just like in the issue report. 

As this did not happen for MSVS 2022 build I used originally, there really is a difference here between MSVC and MinGW. I have what causes it and I run out of time to look into it further (for now?).

As a workaround, I suggest adding an application manifest to your application. This has been basically mandatory since introduction of Windows XP, i.e., at least for two decades. Without it, your application will have 1990s look (very noticeable e.g. for buttons).

**Comment by msbroadf** [comment]
Yes! i was leaving out the wx/msw/wx.rc file from my build accidentally, ill close this ticket...

**Comment by PBfordev** [comment]
@vadz, what about the manifests being embedded in the MSVC and not GCC (and clang) DLLs, is that on purpose?

**Comment by vadz** [comment]
Sorry, I didn't notice your edits.

I think we should embed the manifest in the DLL regardless of the compiler it was built with, so this indeed looks like a bug. I'm not really sure why it happens, looking at the code it looks like it should be included and there are explicit tests for `__GNUC__` in `wx.rc` too, so clearly this did work at some time.

I also don't know what has changed since 3.1.5, the only changes were (your own) 77d8926126 (Simplify Windows application manifests, 2021-12-17) and 419f899b3d (Revert changes to the code generating the manifest name in wx.rc, 2021-12-18), but I don't see how it could have broken this.

To summarize, I'm completely lost here because I think that

1. Applications without any manifest use comctl32.dll v5 which doesn't support transparency.
2. We always embed manifest in the DLL.
3. Nothing has changed since 3.1.5.

And it seems like _none_ of the above is true...

**Comment by PBfordev** [comment]
The manifest is missing from wxWidgets 3.1.5 DLLs already, so it could not be caused by my changes in late 2021.

The DLLs likely do not include `wx.rc`: They contain no resources included in that file such as cursor or bitmaps embedded but they have a version info resource (probably from `src\msw\version.rc` used in make/project files) which is not referenced in `wx.rc`.

Perhaps the manifest in MSVC builds comes from these MSVC `#pragma`s in `defs.h`
https://github.com/wxWidgets/wxWidgets/blob/88e92a5f01f0f3813da866164c7e9aad4ab95d03/include/wx/defs.h#L3205-3233
But would not those `#pragma`s conflict with the manifest embedded from the resource file, and for the applications as well?

After writing all the above, I have just noticed that MSVC DLLs built using nmake do not have the manifest embedded either, so maybe it just gets added in MSVS builds (generated by CMake) by default.

**Comment by vadz** [comment]
Yes, you're right, of course, I have no idea why did I decide that the DLLs used `wx.rc` because they clearly don't. But they do use the code referenced above, which was added back in c8d5853155 (Add rcdefs.h and support for automatic manifests on Windows, 2005-11-08), and this also explains why it's only done for MSVC: the other compilers don't have any equivalent of `#pragma comment(linker)` that could be used to do it for them.

So apparently this would be useful for the applications not including `wx.rc` and not using MSVC default manifest neither... I don't know if it's actually a common use case, but I guess it does no harm to do it. We probably should simplify this by using `processorArchitecture="*"` too, just as you did in 77d892612640a23c17debdac579c4afcbb2b4afd but other than that I don't know if we really need to change anything here.

The only thing I still don't understand is how could it work in 3.1.4 without the application manifest when building using gcc. Could you please confirm that I actually understand correctly that it did work in this case and not just misunderstanding something again?

**Comment by PBfordev** [comment]
This is a screenshot from 3.1.5 treectrl sample built with GCC using CMake-generated makefile, the manifest was deleted from the executable. Please notice that the status bar has 3D-effect, the button is not styled, and the text in controls is blurry compared to the caption, confirming the manifest is not present. Yet, there the transparent pixels around the folder and file icons are not rendered black.
![treectrl-sample-315-no-manifest](https://user-images.githubusercontent.com/12495521/165648454-8f43aac1-c79e-4cf9-9b36-66a2b3bac9bd.png)

And this one is from the current master, processed the same way as 3.1.5.
![treectrl-sample-317-no-manifest](https://user-images.githubusercontent.com/12495521/165649108-8fce0773-8304-4917-8a0a-7d83bbf2f834.png)

I also tried 3.1.5 with the "speaker" PNG embedded the same way as in the patched sample I used in the master, to rule out that the difference is due to changes in `wxArtProvider` where file and folder icons come from. Still worked there.
![treectrl-sample-speaker-315-no-manifest](https://user-images.githubusercontent.com/12495521/165650344-2fa766b9-4d56-4349-8901-7d03dd49c50b.png)

**Comment by vadz** [comment]
This is probably due to some assumptions about alpha support under modern systems that use OS version check rather than the (or without the accompanying) check for the themes being used... It's not awfully important, but I wonder if @a-wi might have a clue about where could this have come from.

In any case, I reopen this because there clearly is a regression here, but I might not have time to do anything about it in 3.1.7.

**Comment by a-wi** [comment]
Bisecting shows the issue was introduced in [689ce0a1](https://github.com/wxWidgets/wxWidgets/commit/689ce0a1383e77ac4d5b18f56f3e597743dc1cdb).

**Comment by vadz** [comment]
Thanks for bisecting this!

I still think that we should use alpha if it's supported, i.e. passing `true` to the `mask` ctor argument should not disable the use of alpha. But we clearly should use mask, and not alpha, if the latter is not supported at all.

So logically `m_useMask` should have 3 states instead of the current 2:

- Don't use mask at all, always use alpha.
- Use mask if there is no alpha, but prefer to use alpha if available.
- Always use mask, converting alpha to mask if necessary.

Does this look right?

**Comment by a-wi** [comment]
Indeed it seems that 3-state internal "use mask" selector should do the trick:
1. `mask` in ctor == false => `ILC_MASK` not set and mask converted to alpha.
2. `mask` in ctor == true && commctl32 >= v6 => `ILC_MASK` set and mask converted to alpha.
3. `mask` in ctor == true && commctl32 < v6 => `ILC_MASK` set and alpha converted to mask.

**Comment by vadz** [comment]
This should be fixed by #22400 but I've finally implemented a different, and even simpler, solution there.

I've tested the treectrl sample and the icons using alpha look fine in it now even when not using the manifest, but please test it too and let me know if you see any regressions. TIA!
